_tools/types/main.go

```diff
@@ -0,0 +1,341 @@
+	uastNodeRe    = regexp.MustCompile("uast:[^\"]*")
+
+	pprof      = flag.Bool("pprof", false, "start pprof profiler http endpoing")
+	skipUpdate = flag.Bool("skip", false, "skip git clone or pull")
+)
+
+func main() {
+	flag.Parse()
+
+	if *pprof {
+		pprofAddr := "localhost:6060"
+		fmt.Fprintf(os.Stderr, "running pprof on %s\n", pprofAddr)
+		go func() {
+			if err := http.ListenAndServe(pprofAddr, nil); err != nil {
+				log.Fatal("cannot start pprof: %v", err)
+			}
+		}()
+	}
+
+	if err := run(); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func run() error {
+	drivers, err := listDrivers()
+	if err != nil {
+		return fmt.Errorf("failed to list drivers: %s", err)
+	}
+
+	if !*skipUpdate {
+		err = maybeCloneOrPullAll(drivers)
+		if err != nil {
+			return fmt.Errorf("failed to pull driver repos: %s", err)
+		}
+	}
+
+	uastTypes := findUASTTypesInSDK()
+	for _, driver := range drivers {
+		if err := analyzeFixtures(driver); err != nil {
+			return err
+		}
+		analyzeCode(driver, uastTypes)
+	}
+
+	formatMarkdownTable(drivers, uastTypes)
+	return nil
+}
+
+type driverStats struct {
+	url          string
+	language     string
+	path         string
+	skip         bool
+	fixturesNum  int
+	fixturesUast map[string]int
+	codeUast     map[string]int
+}
+
+// listDrivers lists all available drivers.
+func listDrivers() ([]*driverStats, error) {
+	fmt.Fprintf(os.Stderr, "discovering all available drivers\n")
+	langs, err := discovery.OfficialDrivers(context.TODO(), &discovery.Options{
+		NoStatic:      true,
+		NoMaintainers: true,
+		NoSDKVersion:  true,
+		NoBuildInfo:   true,
+	})
+	if err != nil {
+		return nil, err
+	}
+	drivers := make([]*driverStats, 0, len(langs))
+	for _, l := range langs {
+		drivers = append(drivers, &driverStats{
+			language:     l.Language,
+			url:          l.RepositoryURL(),
+			path:         l.RepositoryURL()[strings.LastIndex(l.RepositoryURL(), "/"):],
+			fixturesUast: make(map[string]int),
+			codeUast:     make(map[string]int),
+		})
+	}
+	fmt.Fprintf(os.Stderr, "%d drivers available on-line\n", len(langs))
+	return drivers, nil
+}
+
+// maybeCloneOrPullAll either clones repos to path in local FS or, if already preset,
+// pulls the latest master for each of them.
+func maybeCloneOrPullAll(drivers []*driverStats) error {
+	fmt.Fprintf(os.Stderr, "cloning %d drivers to %s\n", len(drivers), reposRootPath)
+	err := os.MkdirAll(reposRootPath, os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	var (
+		pf = make(chan int, 3) // fetches to run in parallel
+		wg sync.WaitGroup
+	)
+	for i := range drivers {
+		wg.Add(1)
+		go func(driver *driverStats) {
+			defer wg.Done()
+			maybeCloneOrPullThrottled(driver, pf)
+		}(drivers[i])
+	}
+	wg.Wait()
+	return nil
+}
+
+// maybeCloneOrPullThrottled is concurent-friendly verison of maybeCloneOrPull.
+// It "throttles" git operarions, based on given chanel capacity.
+func maybeCloneOrPullThrottled(d *driverStats, concurent chan int) error {
```

Comment:
I recommend against plumbing the concurrency down through the helpers. It's helpful for the reader to see all the concurrency matters in one place:

```go
throttle := make(chan struct{}, 3)
var wg sync.WaitGroup
for _, d := range drivers {
  d := d
  wg.Add(1)
  go func() {
     throttle <- struct{}{}
     defer func() { <-throttle ; wg.Done() }()
     maybeCloneOrPull(d)
   }()
}
```

Reply:
Although I see the point, after 74 comments in review I'm starting to have either a self-doubt or a strange feeling that if we would have more people on the team reviewing this code - we would have even more opinions about how to write it, and thus will continue rewriting this simple, non-performance critical, trow-away script 😆